src/fu-util.c

```diff
@@ -560,6 +617,227 @@ fu_util_install_prepared (FuUtilPrivate *priv, gchar **values, GError **error)
 	return TRUE;
 }
 
+static gboolean
+fu_util_history_clear (FuUtilPrivate *priv, gchar **values, GError **error)
+{
+	return fu_pending_remove_all (priv->pending, error);
+}
+
+static gboolean
+fu_util_history_report (FuUtilPrivate *priv, gchar **values, GError **error)
+{
+	guint status_code;
+	g_autofree gchar *config_fn = NULL;
+	g_autofree gchar *data = NULL;
+	g_autofree gchar *machine_hash = NULL;
+	g_autofree gchar *report_uri = NULL;
+	g_autoptr(GKeyFile) kf = g_key_file_new ();
+	g_autoptr(GPtrArray) results = NULL;
+	g_autoptr(JsonBuilder) builder = NULL;
+	g_autoptr(JsonGenerator) json_generator = NULL;
+	g_autoptr(JsonNode) json_root = NULL;
+	g_autoptr(SoupMessage) msg = NULL;
+
+	/* set up networking */
+	if (!fu_util_setup_networking (priv, error))
+		return FALSE;
+
+	/* get the location of the reports */
+	config_fn = g_build_filename (SYSCONFDIR, "fwupd", "client.conf", NULL);
+	if (!g_key_file_load_from_file (kf, config_fn, G_KEY_FILE_NONE, error)) {
+		g_prefix_error (error, "Failed to load %s: ", config_fn);
+		return FALSE;
+	}
+
+	/* get report URI */
+	report_uri = g_key_file_get_string (kf, "fwupd", "ReportURI", NULL);
+	if (report_uri == NULL || report_uri[0] == '\0') {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "ReportURI unset in %s",
+			     config_fn);
+		return FALSE;
+	}
+
+	/* get all updates */
+	results = fu_pending_get_devices (priv->pending, error);
+	if (results == NULL)
+		return FALSE;
+
+	/* get a hash that represents the machine */
+	machine_hash = fu_common_get_machine_hash ("fwupd", error);
+	if (machine_hash == NULL)
+		return FALSE;
+
+	// FIXME: ideally I want to share this code with gnome-software, so we can do one of three things:
+	// 1. copy and paste it
+	// 2. create a GetReportAsJSON dbus method
+	// 3. create a GetHistory dbus method that can marshal the FuDevice additions (etc FirmwareId) as metadata on FwupdDevice
+	// 4. move fu-pending.c to fwupd-history.c and export the API in libfwupd, creating fwupd_history_get_unreported_as_json()
```

Comment:
I like the 1st and 4th option the most.  Plopping it into metadata feels uglier.